XenAdmin/Dialogs/AttachUsbDialog.cs

```diff
@@ -56,16 +56,37 @@ private void BuildList()
             treeUsbList.BeginUpdate();
             try
             {
-                Host[] hosts = _vm.Connection.Cache.Hosts;
-                foreach (Host host in hosts)
+                VM.HA_Restart_Priority SelectedPriority = _vm.HARestartPriority();
+                Pool pool = Helpers.GetPool(_vm.Connection);
+                // Check if HA was enabled on pool and Restart priority was set on VM.
+                bool haEnabled = (pool != null &&
+                    pool.ha_enabled &&
+                    VM.HaPriorityIsRestart(_vm.Connection, SelectedPriority));
+                List<XenRef<Host>> possibleHostRefs = VM.get_possible_hosts(_vm.Connection.Session, _vm.opaque_ref);
+                List<Host> possibleHosts = new List<Host>();
+                foreach (XenRef<Host> possibleHostRef in possibleHostRefs)

```

Comment:
`ResolveAll` can be used here:
`List<Host> possibleHosts = _vm.Connection.ResolveAll(possibleHostRef)`

Reply:
Efficient code! Accepted.